### ISSUE TITLE
How to add negative samples for training

### ISSUE THREAD

**Issue by monkeycc** [ultralytics/ultralytics#5044]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How to add negative samples for training

Dataset format
What should it be like

I didn't see how to do the tutorial
Do you want to directly add the background image to the image folder of the training set
Do I need to write the jpg address in txt

**Comment by glenn-jocher** [comment]
@monkeycc to incorporate negative samples into your training set, you would indeed add the images that contain no objects of interest (your "background" images) directly into the image folder for the training set.

Negative samples help improve the ability of the model to correctly identify the background and handle false positives. These samples should not contain any of the objects your model is being trained to detect. This may be useful in circumstances where the model needs to learn not only when to detect an object, but also when not to register a detection.

Please note that these images should be in the same format as your positive samples, namely JPEG, and adhere to the same dimensions for consistency in training.

In the data labeling process, since these negative images do not contain any object of interest, it can be simply represented by an empty text file (no bounding box coordinates). So, for each of your image of negative sample, there should be a corresponding .txt file, with same base file name as the image, but with no content in it.

Lastly, you must still ensure that the address of these negative .jpg files are listed in the train.txt file, so they're included in the training process.

Thanks for using Ultralytics YOLOv8! Let us know if you have any more questions.

**Comment by monkeycc** [comment]
```
labels/train/
OK_1.txt
OK_2.txt

images/train/
background_1.jpg
background_2.jpg
OK_1.jpg
OK_2.jpg
```
NO_1.jpg
NO_2.jpg


Must add txt?



```
labels/train/
background_1.txt
background_2.txt
OK_1.txt
OK_2.txt

images/train/
background_1.jpg
background_2.jpg
OK_1.jpg
OK_2.jpg
```

background_1.txt
background_2.txt
Content is empty



val
Should we also add negative samples ？

**Comment by markdjthomas** [comment]
I've been modifying the DetectionTrainer to use a custom dataloader that produces data on the fly. If I was to include empty "false positive" examples in the dataloader, do you have any suggestions what the dataloader should yield so that that the bounding boxes can stack correctly? For example should a false positive have bbox coords [0, 0, 0, 0]? If so what should the class label be if 0 is not reserved for the background class in YOLO.

Cheers!

**Comment by glenn-jocher** [comment]
@markdjthomas the inclusion of negative samples, also known as "false positives," can indeed help improve the model's ability to handle actual background and reduce false detections. These samples are ones where there are no objects of interest, essentially just background images.

Regarding your query about what the dataloader should yield, here's the approach you'd typically follow:

When loading negative samples (i.e., images with no bounding boxes), your dataloader needs to produce an empty tensor for the bounding boxes and labels. More specifically, the bounding box tensor could have a shape like `[num_samples, 0, 4]` (where `num_samples` is the number of images in your dataset) and the labels tensor could have a shape like `[num_samples, 0]`. 

So, there's no requirement that the bounding box coordinates for false positives need to be [0, 0, 0, 0] as there won't be any boxes in these images. Similarly, you would not assign a class label since there's no object to classify in these empty samples.

One thing to note here is that YOLOv8, as well as other versions of YOLO, do not use a background class. Rather, object detection is treated as a regression problem for the bounding box coordinates and a multi-label classification problem for the object classes. Anything that does not fall within the calculated bounding boxes is automatically considered as background.

Hope this clarifies the concept of using negative samples within your dataloader. Let me know if you have any more questions. Happy training!

**Comment by markdjthomas** [comment]
Thanks for your help @glenn-jocher!
I did try using empty tensors, but when they're stacked to form a batch they're obviously not included in the final tensor. As a result, the dimensions of the images and bboxes don't match. For example if I have a batch size of 2, my image tensor would have shape [2, 3, 640, 640] but if one of those images is empty, the corresponding bbox tensor for the batch will have shape [1, 4]. Depending on how the data is shuffled this will effect the indexing of image/annotation pairs as theres no guarantee that idx=0 is the labelled image.

Any suggestions, or is this not something the model can currently handle?

**Comment by glenn-jocher** [comment]
@markdjthomas I'm glad to hear that you've successfully managed to incorporate the use of empty tensors and adjust your collate function effectively! It's precisely this type of problem-solving and initiative that helps to improve and innovate within the broader YOLO community. 

Remember, the addition of negative samples can greatly enhance your model's ability to correctly identify the background and handle false positives, making it a worthwhile enhancement.

Thank you for sharing your progress with us and for being a part of the YOLOv8 community. Keep up the good work and don't hesitate to reach out if you have any further questions or issues. Happy coding!

**Comment by monkeycc** [comment]
```
Images/train
Labels/train
```

train  Add negative samples


```
Images/val
Labels/val
```

val need to add negative sample ??

@glenn-jocher

**Comment by glenn-jocher** [comment]
Yes, @monkeycc, it's a good practice to add negative samples to both your training and validation sets. Including these samples, whose associated labels consist of empty .txt files—since there are no objects to detect in them—can help improve the robustness of your model by enabling it to better distinguish between the foreground (i.e., objects of interest) and the background.

As in the training set, don't forget to list the addresses of these negative .jpg files in the validation text file, so they'll be included in the validation process. This addition will ensure the model generalizes better and doesn't overfit to your training data. 

Please proceed with this methodology and let us know if you encounter any issues. Happy training!

**Comment by barbaramartina** [comment]
hi, how do you do to add negative example to train a classification model?... there is no labels folder for classification... should I add a new folder named "negative"?

**Comment by glenn-jocher** [comment]
For classification, “negative samples” need to be an explicit class (e.g. `negative`/`background`) since there’s no `labels/`—just place those images in a folder for that class under both `train/` and `val/` (see the [classification dataset format](https://docs.ultralytics.com/tasks/classify/)):

```text
dataset/
  train/
    cat/
    dog/
    negative/
  val/
    cat/
    dog/
    negative/
```
